### PR TITLE
Ft: add lifecycle task expire functional tests

### PR DESCRIPTION
**Pull request by philipyoo**
Initial commit for LifecycleTask tests. Only tests expiration rule and will add tests for each
rule separately.

Solely tests the LifecycleTask class by counting the number of times a bucket or object Kafka
entry is made and tracks each object entry name.

### CODE REVIEW

File: tests/functional/lifecycle/LifecycleTask.js [hunk 1]
```diff
@@ -0,0 +1,770 @@
...
+            return cb(null, res);
+        }
+
+        let markerIdx = 0;
+        let maxIdx = params.MaxKeys;
+        if (params.KeyMarker) {
+            markerIdx = bucketVersions.findIndex(v =>
+                v.KeyMarker === params.KeyMarker &&
+                v.VersionId === params.VersionIdMarker);
+            if (markerIdx === -1) {
```
Comment (philipyoo) [inline tests/functional/lifecycle/LifecycleTask.js]: Just to note, these errors here were for myself to validate that I correctly implemented the mock functions. I can remove these if requested 👍

**Comment by philipyoo** [comment]
Sorry for any that are currently reviewing. I am planning on removing versioning and mpu functions for now and will add them back in as I add tests for them.

File: tests/functional/lifecycle/LifecycleTask.js [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+PAST.setDate(PAST.getDate() - 5);
+// 5 days after currentDate
+const FUTURE = new Date(CURRENT);
+FUTURE.setDate(FUTURE.getDate() + 5);
+
+const OWNER = 'testOwner';
+
+// NOTE: Since all S3 calls are only 'GET's, I decided to just go for mock
+//  S3 that defines all buckets and their conditions
+class S3Mock {
```
Comment (jonathan-gramain) [inline tests/functional/lifecycle/LifecycleTask.js]: As a general comment, mocking is most useful when testing corner cases or unusual cases (like error injection). When it comes to emulating a real S3 behavior, it can be a better test to use the actual S3 process through functional tests, since creating mocks involves the risk of testing components which do not reflect the behavior of the real ones, as well as increasing complexity of writing and maintaining the mock code itself.

Comment (philipyoo) [reply tests/functional/lifecycle/LifecycleTask.js]: I think as I was writing this, I realized how bulky it was getting and I also was thinking it'd be hard to maintain and extend :/

Will look into using S3 👍

**Comment by philipyoo** [comment]
closing this for now, will open again when all functional tests are ready :)
